core/src/test/java/org/apache/calcite/sql2rel/RelDecorrelatorTest.java

```diff
@@ -86,4 +99,75 @@ public static Frameworks.ConfigBuilder config() {
+      final SqlNode validate = planner.validate(parse);
+      final RelNode originalRel = planner.rel(validate).rel;
+      final HepProgram hepProgram = HepProgram.builder()
+          .addRuleCollection(
+              ImmutableList.of(
+                  // SubQuery program rules
+                  CoreRules.FILTER_SUB_QUERY_TO_CORRELATE,
+                  CoreRules.PROJECT_SUB_QUERY_TO_CORRELATE,
+                  CoreRules.JOIN_SUB_QUERY_TO_CORRELATE,
+                  // plus FilterAggregateTransposeRule
+                  CoreRules.FILTER_AGGREGATE_TRANSPOSE))
+          .build();
+      final Program program =
+          Programs.of(hepProgram, true, Objects.requireNonNull(cluster.getMetadataProvider()));
+      final RelNode before =
+          program.run(cluster.getPlanner(), originalRel, cluster.traitSet(),
+              Collections.emptyList(), Collections.emptyList());
+      final String planBefore = ""
+          + "LogicalProject(EXPR$0=[1])\n"
+          + "  LogicalProject(DEPTNO=[$0], TOTAL=[$1])\n"
+          + "    LogicalFilter(condition=[>($1, $2)])\n"
+          + "      LogicalCorrelate(correlation=[$cor0], joinType=[left], requiredColumns=[{0}])\n"
+          + "        LogicalAggregate(group=[{0}], TOTAL=[SUM($1)])\n"
+          + "          LogicalProject(DEPTNO=[$7], SAL=[$5])\n"
+          + "            LogicalTableScan(table=[[scott, EMP]])\n"
+          + "        LogicalAggregate(group=[{}], EXPR$0=[AVG($0)])\n"
+          + "          LogicalProject(TOTAL=[$1])\n"
+          + "            LogicalAggregate(group=[{0}], TOTAL=[SUM($1)])\n"
+          + "              LogicalFilter(condition=[=($cor0.DEPTNO, $0)])\n"
+          + "                LogicalProject(DEPTNO=[$7], SAL=[$5])\n"
+          + "                  LogicalTableScan(table=[[scott, EMP]])\n";
+      assertThat(before, hasTree(planBefore));
+
+      // Check decorrelation does not fail here
+      final RelNode after = RelDecorrelator.decorrelateQuery(before, builder);
+
+      // Verify plan
+      final String planAfter = ""
+          + "LogicalProject(EXPR$0=[1])\n"
+          + "  LogicalJoin(condition=[AND(=($0, $2), >($1, $3))], joinType=[inner])\n"
+          + "    LogicalAggregate(group=[{0}], TOTAL=[SUM($1)])\n"
+          + "      LogicalProject(DEPTNO=[$7], SAL=[$5])\n"
+          + "        LogicalTableScan(table=[[scott, EMP]])\n"
+          + "    LogicalAggregate(group=[{0}], EXPR$0=[AVG($1)])\n"
+          + "      LogicalProject(DEPTNO=[$0], TOTAL=[$1])\n"
+          + "        LogicalAggregate(group=[{0}], TOTAL=[SUM($1)])\n"
+          + "          LogicalProject(DEPTNO=[$0], SAL=[$1])\n"
+          + "            LogicalFilter(condition=[IS NOT NULL($0)])\n"
+          + "              LogicalProject(DEPTNO=[$7], SAL=[$5])\n"
+          + "                LogicalTableScan(table=[[scott, EMP]])\n";
+      assertThat(after, hasTree(planAfter));
+    } catch (Exception e) {
+      throw TestUtil.rethrow(e);
```

Comment:
why do you have to catch and rethrow?

Reply:
It was just because of the exceptions declared by parse and validation. I have isolated the try-catch to the place where it is required, to make things more clear.